text/0006-sampling.md

```diff
@@ -0,0 +1,300 @@
+# Sampling API
+
+*Status: proposed*
+
+## TL;DR
+This section tries to summarize all the changes proposed in this RFC:
+ 1. Move the `Sampler` interface from the API to SDK package. Apply some minor changes to the
+ `Sampler` API.
+ 1. Add a new `SamplerHint` concept to the API package.
+ 1. Add capability to record `Attributes` that can be used for sampling decision during the `Span`
+ creation time.
+ 1. Add capability to start building a `Span` with a delayed `build` method. This is useful for
+ cases where some `Attributes` that are useful for sampling are not available when start building
+ the `Span`. As an example in Java the current `Span.Builder` will use as a start time for the
+ `Span` the moment when the builder is created and not the moment when the `build()` method is
+ called.
+ 1. Remove `addLink` APIs from the `Span` interface, and allow recording links only during the span
```

Comment:
Is this required? If I wanted to record my children in the parent span, a very natural way to do that would be via adding a link. In the case of attributes, we now have an API that lets you provide them at start time or during the span's lifecycle. Could we do the same with links?  This change doesn't feel like a requirement, considering how attributes are being treated.

Reply:
I think we need to revisit links. Currently it has a lot of unclear stuff:
1. What's the maximum allowed number of links?
2. Do we in general keep links to all the children in a _logical parent_ span, or we put `logical parent` as a link in the child span? (or both?)
3. Do we allow links to be added postmortem (after both source and target spans are exported)?

We should track this in a separate PR instead of blocking this one.

Reply:
@jmacd I applied a simple rule when I suggested this: adding a new API is considered backwards compatible so delaying this decision seems to be fine. this is why I proposed this :)

Also if we do apply the links later is going to be a discussion if they can influence the sampling/recording decision.

Reply:
Closing this conversation, please re-open if needed.